File: modules/core/src/main/scala/com/github/chengpohi/dsl/eql/Definition.scala
package com.github.chengpohi.dsl.eql

import cats.implicits._
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization.write

import scala.concurrent.Future


trait Definition[A] {
  def execute: Future[A]

  def extract(path: String): ExtractDefinition = {
    ExtractDefinition(this, path)
  }

  def json: String
}


case class ExtractDefinition(definition: Definition[_], path: String)
  extends Definition[String] with FutureOps {


  override def execute: Future[String] = {
    val jObj = parse(definition.json)
    val result = path.split("\\.").foldLeft(jObj) { (o, i) =>
      o \ i
    }
    write(result).pure[Future]
  }

  override def json: String = execute.await
}


case class ErrorHealthRequestDefinition(error: String)
  extends Definition[String] {
  override def execute: Future[String] = Future.failed(new RuntimeException(error))

  override def json: String = error
}


case class PureStringDefinition(s: String) extends Definition[String] {
  override def execute: Future[String] = Future.successful(s)

  override def json: String = s
}

File: src/main/java/com/github/chengpohi/context/Definition.kt
package com.github.chengpohi.context

import com.github.chengpohi.script.ScriptContext
import kotlinx.coroutines.future.await
import java.util.concurrent.CompletableFuture

interface Definition<A> {
    suspend fun execute(context: ScriptContext): A
    
    fun json(context: ScriptContext): String
}

data class ErrorHealthRequestDefinition(val error: String) : Definition<String> {
    override suspend fun execute(context: ScriptContext): String {
        throw RuntimeException(error)
    }
    
    override fun json(context: ScriptContext): String = error
}

data class PureStringDefinition(val s: String) : Definition<String> {
    override suspend fun execute(context: ScriptContext): String = s
    
    override fun json(context: ScriptContext): String = s
} 